nixos/modules/services/backup/borgmatic.nix

```diff
@@ -0,0 +1,247 @@
+        '';
+        example = "3H";
+      };
+
+      keep_secondly = mkOption {
+        type = with types; nullOr int;
+        default = null;
+        description = ''
+          Number of secondly archives to keep.
+        '';
+        example = 60;
+      };
+
+      keep_minutely = mkOption {
+        type = with types; nullOr int;
+        default = null;
+        description = ''
+          Number of minutely archives to keep.
+        '';
+        example = 60;
+      };
+
+      keep_hourly = mkOption {
+        type = with types; nullOr int;
+        default = null;
+        description = ''
+          Number of hourly archives to keep.
+        '';
+        example = 24;
+      };
+
+      keep_daily = mkOption {
+        type = with types; nullOr int;
+        default = null;
+        description = ''
+          Number of daily archives to keep.
+        '';
+        example = 7;
+      };
+
+      keep_weekly = mkOption {
+        type = with types; nullOr int;
+        default = null;
+        description = ''
+          Number of weekly archives to keep.
+        '';
+        example = 4;
+      };
+
+      keep_monthly = mkOption {
+        type = with types; nullOr int;
+        default = null;
+        description = ''
+          Number of monthly archives to keep.
+        '';
+        example = 6;
+      };
+
+      keep_yearly = mkOption {
+        type = with types; nullOr int;
+        default = null;
+        description = ''
+          Number of yearly archives to keep.
+        '';
+        example = 1;
+      };
+
+      prefix = mkOption {
+        type = with types; nullOr str;
+        default = null;
+        description = ''
+          When pruning, only consider archive names starting with this
+          prefix.  Borg placeholders can be used. See the output of
+          "borg help placeholders" for details. Defaults to
+          "{hostname}-". Use an empty value to disable the default.
+        '';
+        example = "sourcehostname";
+      };
+
+    };
+
+    extraConfig = mkOption {
+      type = with types; attrsOf anything;
+      default = { };
+      description = ''
+        Additional configuration to append to.
+        See https://torsion.org/borgmatic/docs/reference/configuration/
+      '';
+      example = literalExample ''
+        {
+          consistency = {
+            checks = [
+              "repository"
+              "archives"
+            ];
+          };
+          hooks = {
+            before_backup = [ "prepare-for-backup.sh" ];
+            postgresql_databases = [
+              {
+                name = "users"
+              }
+            ];
+            healthchecks = "https://hc-ping.com/be067061-cf96-4412-8eae-62b0c50d6a8c";
+          };
+        }
+      '';
+    };
+
+  };
+
+  config = mkIf cfg.enable {
+
+    environment.systemPackages = [ pkgs.borgmatic ];
+
+    environment.etc."borgmatic/config.yaml".source = cfgfile;
+
+    systemd.services.borgmatic = {
```

Comment:
Given your unit is mostly a copy+paste of the [upstream unit](https://github.com/witten/borgmatic/blob/master/sample/systemd/borgmatic.service) I wonder if you're better off just utilizing the upstream unit from the package, via `systemd.packages = [ pkgs.borgmatic ];`and then adding any customization as needed... :thinking:

Reply:
I'm not going to relitigate this because it'd be awful for the contributing process, but IMO it'd have been better to keep the manually-handled service and replace the `environment.etc` usage by the `-c` command-line flag to borgmatic: the fewer things in `/etc`, the more self-contained the modules are, the more hope we can have for things like monadic modules.

Now well it's something that'll probably not come for a while so let's just land, but I wanted to say this before, so we can discuss it (here or on #nixos-dev maybe?) :)

Reply:
As discussed in #nixos-dev we can use the upstream units via `systemd.packages` and override the `ExecStart` line to provide a custom path (ie. something in the nix store) for configuration files. This allows us to use upstream units all while keeping configuration out of `/etc`, keeping the hope of monadic modules alive.